x-pack/plugins/security_solution/public/resolver/store/data/selectors.ts

```diff
@@ -298,44 +222,65 @@ export const relatedEventDisplayInfoByEntityAndSelfID: (
+) => (node: string, eventCategory: string) => SafeResolverEvent[] = createSelector(
   relatedEventsByEntityId,
   function (
     /* eslint-disable no-shadow */
     relatedEventsByEntityId
     /* eslint-enable no-shadow */
   ) {
-    return defaultMemoize((entityId: string) => {
-      return defaultMemoize((ecsCategory: string) => {
-        const relatedById = relatedEventsByEntityId.get(entityId);
-        // With no related events, we can't return related by category
-        if (!relatedById) {
-          return [];
+    // A map of nodeID -> event category -> SafeResolverEvent[]
+    const nodeMap: Map<string, Map<string, SafeResolverEvent[]>> = new Map();
+    for (const [nodeID, events] of relatedEventsByEntityId) {
+      // A map of eventCategory -> SafeResolverEvent[]
+      let categoryMap = nodeMap.get(nodeID);
+      if (!categoryMap) {
+        categoryMap = new Map();
+        nodeMap.set(nodeID, categoryMap);
+      }
+
+      for (const event of events.events) {
+        for (const category of eventModel.eventCategory(event)) {
+          let eventsInCategory = categoryMap.get(category);
+          if (!eventsInCategory) {
+            eventsInCategory = [];
+            categoryMap.set(category, eventsInCategory);
+          }
+          eventsInCategory.push(event);
         }
-        return relatedById.events.reduce(
-          (eventsByCategory: ResolverEvent[], candidate: ResolverEvent) => {
-            if (
-              [candidate && eventModel.allEventCategories(candidate)].flat().includes(ecsCategory)
-            ) {
-              eventsByCategory.push(candidate);
-            }
-            return eventsByCategory;
-          },
-          []
-        );
-      });
-    });
+      }
+    }
+
+    // Use the same empty array for all values that are missing
+    const emptyArray: SafeResolverEvent[] = [];
+
+    return (entityID: string, category: string): SafeResolverEvent[] => {
+      const categoryMap = nodeMap.get(entityID);
+      if (!categoryMap) {
+        return emptyArray;
+      }
+      const eventsInCategory = categoryMap.get(category);
+      return eventsInCategory ?? emptyArray;
+    };
   }
 );
 
-/**
- * returns a map of entity_ids to booleans indicating if it is waiting on related event
- * A value of `undefined` can be interpreted as `not yet requested`
- * @deprecated
- */
-export function relatedEventsReady(data: DataState): Map<string, boolean> {
-  return data.relatedEventsReady;
-}
+export const relatedEventCountByType: (
+  state: DataState
+) => (nodeID: string, eventType: string) => number | undefined = createSelector(
+  relatedEventsStats,
+  (statsMap) => {
+    return (nodeID: string, eventType: string): number | undefined => {
+      const stats = statsMap(nodeID);
+      if (stats) {
+        const value = Object.prototype.hasOwnProperty.call(stats.events.byCategory, eventType);
+        if (typeof value === 'number') {
```

Comment:
:grey_question: Maybe for extra safety like `Number.isFinite(value)` so we don't pass `NaN`s?